src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/AbstractMoveCommand.cs

```diff
@@ -41,76 +41,40 @@ public AbstractMoveCommand(IPhysicalProjectTree projectTree, SVsServiceProvider
 
-                    // If we moved a folder, collapse it.
-                    if (node.IsFolder)
-                    {
-                        window.ExpandItem(hierarchy, itemId, EXPANDFLAGS.EXPF_CollapseFolder);
-                    }
-                }
+                // Deliberately not async to reduce the type of
+                // code you can run while holding the lock.
+                return action(evaluatedProject);
             }
-
-            return didMove;
         }
 
-        /// <summary>
-        /// Get reference to IVsUIHierarchyWindow interface from guid persistence slot.
-        /// </summary>
-        /// <param name="serviceProvider">The service provider.</param>
-        /// <param name="persistenceSlot">Unique identifier for a tool window created using IVsUIShell::CreateToolWindow.
-        /// The caller of this method can use predefined identifiers that map to tool windows if those tool windows
-        /// are known to the caller. </param>
-        /// <returns>A reference to an IVsUIHierarchyWindow interface, or <c>null</c> if the window isn't available, such as command line mode.</returns>
-        private static IVsUIHierarchyWindow GetUIHierarchyWindow(IServiceProvider serviceProvider, Guid persistenceSlot)
+        protected override async Task<bool> TryHandleCommandAsync(IProjectTree node, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut)
         {
-            if (serviceProvider == null)
-            {
-                throw new ArgumentNullException(nameof(serviceProvider));
-            }
+            var didMove = await OpenProjectForWriteAsync(_configuredProject, project => TryMove(project, node)).ConfigureAwait(true);
 
-            IVsUIShell shell = serviceProvider.GetService(typeof(SVsUIShell)) as IVsUIShell;
-
-            if (shell == null)
-            {
-                throw new InvalidOperationException();
-            }
-
-            object pvar = null;
-            IVsUIHierarchyWindow uiHierarchyWindow = null;
-
-            try
-            {
-                if (ErrorHandler.Succeeded(shell.FindToolWindow(0, ref persistenceSlot, out var frame)) && frame != null)
-                {
-                    ErrorHandler.ThrowOnFailure(frame.GetProperty((int)__VSFPROPID.VSFPROPID_DocView, out pvar));
-                }
-            }
-            finally
+            if (didMove)
             {
-                if (pvar != null)
-                {
-                    uiHierarchyWindow = (IVsUIHierarchyWindow)pvar;
-                }
+                // Wait for updating to finish before re-selecting the node that moved.
+                // We need to re-select the node after it is moved in order to continuously move the node using hotkeys.
+                await _projectTree.TreeService.PublishLatestTreeAsync(waitForFileSystemUpdates: true).ConfigureAwait(true);
+                HACK_NodeHelper.Select(_configuredProject, _serviceProvider, node);
```

Comment:
This looks like it needs the UI thread. You should have a switch to UI thread. Don't rely on the fact that commands are _typically_ called on the UI thread.

Reply:
Yes, I've addressed this in the other PR.